velox/functions/lib/Re2Functions.cpp

```diff
@@ -970,51 +969,229 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> re2ExtractSignatures() {
+      charKind_ = CharKind::kEscape;
+    } else if (
+        previousCharKind_ != CharKind::kEscape && currentChar != escapeChar_ &&
+        (currentChar == '_' || currentChar == '%')) {
+      charKind_ = CharKind::kWildcard;
+    } else {
+      charKind_ = CharKind::kNormal;
+    }
+  }
+
+  /// Current index of the cursor.
+  char currentIndex() const {
+    return currentIndex_;
+  }
+
+  /// Char at current cursor.
+  char current() const {
+    return pattern_.data()[currentIndex_];
+  }
+
+  CharKind state() const {
+    return charKind_;
+  }
+
+  CharKind previousState() const {
+    return previousCharKind_;
+  }
+
+ private:
+  const StringView pattern_;
+  const std::optional<char> escapeChar_;
+
+  int32_t currentIndex_ = -1;
+  // State of current char.
+  CharKind charKind_ = CharKind::kNormal;
+  // State of previous char.
+  CharKind previousCharKind_ = CharKind::kNormal;
+};
+
+PatternMetadata determinePatternKind(
+    StringView pattern,
+    std::optional<char> escapeChar) {
+  int32_t patternLength = pattern.size();
+
+  // Index of the first % or _ character(not escaped).
+  int32_t wildcardStart = -1;
   // Index of the first character that is not % and not _.
-  vector_size_t fixedPatternStart = -1;
+  int32_t fixedPatternStart = -1;
   // Index of the last character in the fixed pattern, used to retrieve the
   // fixed string for patterns of type kSubstring.
-  vector_size_t fixedPatternEnd = 0;
+  int32_t fixedPatternEnd = -1;
+  // Count of wildcard character sequences in pattern.
+  size_t numWildcardSequences = 0;
   // Total number of % characters.
-  vector_size_t anyCharacterWildcardCount = 0;
+  size_t anyCharacterWildcardCount = 0;
   // Total number of _ characters.
-  vector_size_t singleCharacterWildcardCount = 0;
-  auto patternStr = pattern.data();
+  size_t singleCharacterWildcardCount = 0;
+
+  PatternStringIterator iterator{pattern, escapeChar};
+  bool fallbackToGeneric = false;
+
+  // Return false if failed to handle the new fixed pattern(need to fallback),
+  // true otherwise.
+  auto handleNewFixedPattern = [&]() -> bool {
+    // Record the first fixed pattern start.
+    if (fixedPatternStart == -1) {
+      fixedPatternStart = iterator.currentIndex();
+    } else {
+      // This is not the first fixed pattern, not supported, so fallback.
+      if (iterator.previousState() ==
+          PatternStringIterator::CharKind::kWildcard) {
+        fallbackToGeneric = true;
+        return false;
+      }
+    }
+
+    return true;
+  };
 
-  while (i < patternLength) {
-    if (patternStr[i] == '%' || patternStr[i] == '_') {
-      if (wildcardStart == -1) {
-        wildcardStart = i;
+  // Iterator through the pattern string to collect the stats for the simple
+  // patterns that we can optimize: e.g. the
+  while (iterator.hasNext()) {
+    iterator.next();
+    auto current = iterator.current();
+
+    if (iterator.previousState() == PatternStringIterator::CharKind::kEscape) {
```

Comment:
The iterator abstraction is supposed to hide the details of how escape characters are processed. I'd expect it to allow us to write code as if we don't know anything about escape characters. However, I see that we are handling escape chars both in the iterator and in the code that uses the iterator. Is this intentional?

Reply:
In my initial design, `PatternStringIterator` does not encapsulate the `escape` state, but I do agree with you that let  it encapsulate `escape` state will make its user much easier.